### PR TITLE
fix(dub): Make config.d executable to incremental compilation

### PR DESCRIPTION
**Pull request by ljmf00**
Signed-off-by: Luís Ferreira <contact@lsferreira.net>

---

Before:
```
Running pre-generate commands for dmd:lexer...
parsePackageRecipe dub.sdl
Performing "debug" build using /usr/bin/dmd for x86_64.
config ~master: building configuration "application"...
Linking...
Running generated/dub/config /home/luis/Workspace/Programming/Repos/temp/dmdlint/dmd/generated/dub /home/luis/Workspace/Programming/Repos/temp/dmdlint/dmd/VERSION /etc
Performing "debug" build using /usr/bin/dmd for x86_64.
dmd:root ~master: target for configuration "library" is up to date.
dmd:lexer ~master: building configuration "library"...
dmd:parser ~master: building configuration "library"...
dmd:frontend ~master: building configuration "library"...
```

After:
```
Running pre-generate commands for dmd:lexer...
Performing "debug" build using /usr/bin/dmd for x86_64.
dmd:root ~master: target for configuration "library" is up to date.
dmd:lexer ~master: target for configuration "library" is up to date.
dmd:parser ~master: target for configuration "library" is up to date.
dmd:frontend ~master: target for configuration "library" is up to date.
```

### CODE REVIEW

File: config.d [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#!/usr/bin/env dub
```
Comment (ljmf00) [inline config.d:1]: Make this executable, in case someone wants to run it as a script on POSIX systems.

File: dub.sdl [hunk 2]
```diff
@@ -35,10 +35,7 @@ subPackage {
     "DMDLIB"
 
   preGenerateCommands `
-    "$${DUB_EXE}" \
-    --arch=$${DUB_ARCH} \
-    --compiler=$${DC} \
-    --single "$${DUB_PACKAGE_DIR}config.d" \
+    "$${DUB_PACKAGE_DIR}config.d" \
```
Comment (ljmf00) [inline dub.sdl:38]: For some reason, this dub call doesn't incrementally compile, but calling it directly does.

**Comment by ljmf00** [comment]
@CyberShadow Is this infrastructure open? It should provide the full absolute compiler path in order to work, instead of relying on paths relative to cwd.

**Comment by CyberShadow** [comment]
This can't possibly work on Windows, right? It has no notion of shebangs or executable files.

> @CyberShadow Is this infrastructure open?

Yes: https://github.com/cybershadow/ae/blob/master/sys/d/manager.d

> It should provide the full absolute compiler path in order to work, instead of relying on paths relative to cwd.

Not sure if it would be fair to impose such a requirement on everyone.

**Comment by ljmf00** [comment]
> This can't possibly work on Windows, right? It has no notion of shebangs or executable files.

It works as the compiler safely ignores the shebang. Also, for Windows, a dub call is used. I'm only changing the way is called for POSIX platforms, but see bellow. Anyway, `dub` should fix those issues but I'm not willing to fix them due to how messy `dub` is.

> > @CyberShadow Is this infrastructure open?
> 
> Yes: [cybershadow/ae@`master`/sys/d/manager.d](https://github.com/cybershadow/ae/blob/master/sys/d/manager.d)
> 
> > It should provide the full absolute compiler path in order to work, instead of relying on paths relative to cwd.
> 
> Not sure if it would be fair to impose such a requirement on everyone.

Right, that is a problem, and probably `dub` shouldn't produce this behaviour, but given that rationale, I will revert this.

**Comment by CyberShadow** [comment]
> It works as the compiler safely ignores the shebang. Also, for Windows, a dub call is used.

Ah yes, I see the `platform="posix"` now, thanks for the clarification. :+1:

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
How does this make config.d do incremental compilation?

**Comment by ljmf00** [comment]
> How does this make config.d do incremental compilation?

The generated folder for config binary should be in a different folder, other than the import paths. If it is the same, dub detect changes and recompile it.

**Pull request merged**